app/functoria_app.ml

```diff
@@ -622,36 +622,53 @@ module Make (P: S) = struct
+    and file_ml = Fpath.basename file_ml
     in
     Bos.OS.Path.matches Fpath.(config_dir / "_build" // cfg + "$(ext)") >>= fun files ->
     List.fold_left (fun r p -> r >>= fun () -> Bos.OS.Path.delete p) (R.ok ()) files >>= fun () ->
     with_current config_dir
       (fun () ->
-         let pkgs = match P.packages with
-           | []   -> Bos.Cmd.empty
-           | pkgs ->
-             let pkgs =
-               List.fold_left (fun acc pkg ->
-                   String.Set.union pkg.ocamlfind acc
-                 ) String.Set.empty pkgs
-               |> String.Set.elements
-             in
-             Bos.Cmd.(v "-pkgs" % String.concat ~sep:"," pkgs)
-         in
-         let ignore_dirs = match P.ignore_dirs with
-           | []    -> Bos.Cmd.empty
-           | dirs  -> Bos.Cmd.(v "-Xs" % String.concat ~sep:"," dirs)
-         in
-         let cmd =
-           Bos.Cmd.(v "ocamlbuild" % "-use-ocamlfind" % "-classic-display" %
-                    "-tags" % "bin_annot" % "-quiet" %%
-                    ignore_dirs %% pkgs % file)
+          let target_dir = Fpath.(v "_build")
+          (* Import files in the build target *)
+          and target_path = Fpath.(v "_build" / file_ml)
+          and target_path_dune = Fpath.(v "_build" / "dune.inc")
+          in
+          Bos.OS.Dir.create target_dir >>= fun _ ->
+          Bos.OS.Path.symlink ~force:true ~target:(Fpath.(v ".." / file_ml)) target_path >>= fun () ->
+          Bos.OS.Path.exists Fpath.(v "dune.inc") >>= fun res ->
+          (match res with
+          | true ->  (Bos.OS.Path.symlink ~force:true ~target:(Fpath.(v ".." / "dune.inc")) target_path_dune >>= fun () -> Ok true)
+          | false -> Ok false)
+          (* Generate dune configuration file *)
+          >>= fun has_dune_inc ->
+          let config_file = Fpath.(v "_build" / "dune") in
+          let pkgs = match P.packages with
+            | []   -> ""
+            | pkgs ->
+              let pkgs =
+                List.fold_left (fun acc pkg ->
+                    String.Set.union pkg.ocamlfind acc
+                  ) String.Set.empty pkgs
+                |> String.Set.elements
+              in
+              String.concat ~sep:" " pkgs
+          in
+          let dune_file = "(library (name config) (modules config) (libraries "^pkgs^"))" in
+          let dune_file = if has_dune_inc then ("(include dune.inc)"^dune_file) else dune_file in
+          let write_dune_file = Bos.OS.File.write config_file dune_file in
+          let write_dune_workspace_file = Bos.OS.File.write Fpath.(v "_build" / "dune-project") "(lang dune 1.7)"
+          in
+          (* Build config.cmxa with dune *)
+          let target_file = Fpath.(v "_build" / "default" / file) |> Fpath.to_string in
+          let cmd =
+            Bos.Cmd.(v "dune" % "build" % "--no-print-directory" % "--root" % "_build" % target_file)
          in
+         write_dune_workspace_file >>= fun _ -> write_dune_file >>= fun _ ->
```

Comment:
not sure what to do with potential errors here -- should maybe the files (`dune-project` / `dune`) be removed if they exist before? (the scenario i have in mind is that i do a mirage configure, modify config.ml (adjusting dependencies) and run it again...)

Reply:
Yes, I fixed that by removing the files if they exist before.